# eBay Car Sales
## Introduction
This project explores ~50,000 data points on used cars posted on eBay Kleinanzeigen, a classifieds section of the German eBay website. The data dictionary included with the dataset is as follows:

| Column name | Description |
|---|---|
| dateCrawled | When this ad was first crawled. All field-values are taken from this date. |
| name | Name of the car. |
| seller | Whether the seller is private or a dealer. |
|offerType | The type of listing |
| price | The price on the ad to sell the car. |
| abtest | Whether the listing is included in an A/B test. |
|vehicleType | The vehicle Type.|
|yearOfRegistration | The year in which the car was first registered.|
|gearbox | The transmission type.|
|powerPS | The power of the car in PS.|
|model | The car model name.|
|kilometer | How many kilometers the car has driven.|
|monthOfRegistration | The month in which the car was first registered.|
|fuelType | What type of fuel the car uses.|
|brand | The brand of the car.|
|notRepairedDamage | If the car has a damage which is not yet repaired.|
|dateCreated | The date on which the eBay listing was created.|
|nrOfPictures | The number of pictures in the ad.|
|postalCode | The postal code for the location of the vehicle.|
|lastSeenOnline | When the crawler saw this ad last online.|

The aim is to analyze the included used car listings

## 1 - Importing and Cleaning Data

In [2]:
import pandas as pd
import numpy as np
autos = pd.read_csv('Datasets/autos.csv', encoding = 'Latin-1')

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Hark data! A couple of things we will probably have to clean up:
* Looks like we have a couple of missing entries in the `vehicleType`,`gearbox`,`model`,`fuelType`, and `notRepairedDamage` columns.
* A lot of the date/time columns are of dtype object instead of datetimes.
* The `price` column is also an object not an int or float. We should probably take care of that.

#### Rename columns
I'll change the names of the columns in the original dataset to match the convention that snakecase is used for variables/functions, and camelcase is reserved for classes. This should hopefully make my work more readable.

In [4]:
new_cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.columns = new_cols
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Now lets get another overview of the data.

In [5]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

The following columns have are good candidates to be dropped because they have low unique value count and most of the data points have the same value:
* `seller`
* `offer_type`
* `gearbox` maybe

The following columns are contain numerical data as non-numerical dtypes (should be converted to numerical):
* `price`
* `odometer`

The following columns need a closer look:
* `registration_year` - seems like the min year of registration is 1000, and the max is 9999. That doesn't really make sense.
* `power_ps` - 17700 ps for a car is unseemly. In fact unrealistic. This is probably erroneous.
* `nr_of_pictures` - seems like all values in this column are zero...which would make the column useless. Worth a second look.

### 1.1 - Columns needing a closer look

Let's investigate some of these columns that need a closer look

In [6]:
autos.loc[(autos['registration_year'] > 2022) | (autos['registration_year'] < 1900), 'registration_year'].value_counts().sum()

24

Seems like there are in fact some erroneous values in the `registration_year` column. There are 24 values in total with registration years before 1900 and after present day (2022). We'll probably want to screen out a lot of these values. Let's examine the `power_ps` column next.

In [7]:
autos.loc[autos['power_ps'] > 1460, 'power_ps'].value_counts().sum()

33

A quick google search shows that the car with the highest power (in ps) is about 1458 ps. With this information, let's say that any car with `power_ps > 1460` is probably erroneous. With this rule, ~33 car entries are erroneous.
Finally, let's look at the `nr_of_pictures` column.

In [8]:
autos['nr_of_pictures'].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

Yep. As suspected, this column is entirely useless since it has only one value. All entries are zero.

### 1.2 - Converting dtypes

First we'll need to convert the `price` and `odometer` columns into numeric dtypes

In [9]:
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(float)
autos['price'].head()

/var/folders/5f/1dm3w4px4b3cvb1gz1gfkx5w0000gn/T/ipykernel_24724/2486208230.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos['price'] = autos['price'].str.replace('$','')


0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
Name: price, dtype: float64

In [10]:
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].astype(int)
autos['odometer'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64

We've turned all numeric data (stored as strings) into numerical dtypes; finally let's change some column names to ensure the information of these columsn isn't lost.

In [11]:
autos.rename(columns = {'odometer':'odometer_km','price':'price_usd'}, inplace = True)

### 1.3 - A closer look at the `price` column

In [12]:
display(autos.describe())
display(autos['price_usd'].value_counts().sort_index(ascending = False).head(20))
display(autos['price_usd'].value_counts().sort_index().head(20))

price_usd  registration_year      power_ps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        50000.000000         50000.0  50000.000000  
mean             5.723360             0.0  50813.627300  
std              3.711984             0.0  25779.747957  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30451.000000  
50%              6.000000             0.0  49577.000000  
75%              9.000000             0.0  71540.000000  
max             12.000000             0.0  99998.000000

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price_usd, dtype: int64

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price_usd, dtype: int64

Upon investigating the `price` column more closely, it looks like some of the prices are outliers. It's hard to believe some of these cars are going for over \\$1,000,000 and some even selling for free on eBay. Let's apply a screen that removes entries with prices less than \\$200 and more than \\$500,000. We'll screen out some of these prices.

In [13]:
autos.loc[autos['price_usd'] < 200, 'price_usd'].value_counts().sort_index().sum()

2341

In [14]:
price_lower_limit = 200
price_upper_limit = 500000
autos = autos.loc[autos['price_usd'].between(price_lower_limit, price_upper_limit)]

### 1.4 - A closer look at `datetime` columns
Now let's move on to the columns with datetime information

In [15]:
autos[['date_crawled','ad_created','last_seen']]
display(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index())
display(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.025354
2016-03-06    0.014062
2016-03-07    0.035995
2016-03-08    0.033120
2016-03-09    0.033036
2016-03-10    0.032322
2016-03-11    0.032700
2016-03-12    0.036877
2016-03-13    0.015699
2016-03-14    0.036562
2016-03-15    0.034232
2016-03-16    0.029447
2016-03-17    0.031546
2016-03-18    0.012824
2016-03-19    0.034610
2016-03-20    0.037800
2016-03-21    0.037360
2016-03-22    0.032700
2016-03-23    0.032385
2016-03-24    0.029321
2016-03-25    0.031420
2016-03-26    0.032217
2016-03-27    0.031210
2016-03-28    0.035030
2016-03-29    0.033980
2016-03-30    0.033897
2016-03-31    0.031840
2016-04-01    0.033813
2016-04-02    0.035681
2016-04-03    0.038724
2016-04-04    0.036562
2016-04-05    0.013139
2016-04-06    0.003169
2016-04-07    0.001364
Name: date_crawled, dtype: float64

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038976
2016-04-04    0.036919
2016-04-05    0.011859
2016-04-06    0.003253
2016-04-07    0.001217
Name: ad_created, Length: 76, dtype: float64

Seems like the bulk of the `date_crawled` values is between early March 2016 and early April 2016. Additionally, it looks like the bulk of the `ad_created` entries happen between early March 2016 and early April 2016. Let's look at the `registration_year` column

In [16]:
display(autos['registration_year'].describe())
display(autos['registration_year'].value_counts().sort_values(ascending=False).head(40).sort_index())
display(autos['registration_year'].value_counts().sort_index().head(10))
display(autos['registration_year'].value_counts().sort_index().tail(15))

count    47645.000000
mean      2004.800084
std         88.423872
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

1970      35
1978      42
1980      78
1982      41
1983      51
1984      51
1985      88
1986      71
1987      72
1988     133
1989     169
1990     320
1991     330
1992     355
1993     406
1994     593
1995    1119
1996    1302
1997    1852
1998    2285
1999    2845
2000    3019
2001    2611
2002    2470
2003    2692
2004    2694
2005    2906
2006    2667
2007    2272
2008    2206
2009    2079
2010    1587
2011    1617
2012    1307
2013     797
2014     656
2015     376
2016    1161
2017    1371
2018     468
Name: registration_year, dtype: int64

1000    1
1001    1
1111    1
1800    2
1910    2
1927    1
1929    1
1931    1
1934    2
1937    4
Name: registration_year, dtype: int64

2015     376
2016    1161
2017    1371
2018     468
2019       1
2800       1
4100       1
4500       1
4800       1
5000       3
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

Seems like the bulk of the `registration_year` data is between the mid nineties and late 2010's. However it seems like we have a couple of outliers: cars were invented in 1885, yet we have cars registered in year 1000. Additionally, we have cars that were registered after the latest `date_crawled` value. Per a cursory google search, it seems like car registrations in Germany are good for, at most, 2 years after their original registration date. It's possible that ad listers interpreted `registration_year` as the year the registration expires, or then year the vehicle was registered. To accommodate for both possibilities, the latest year that makes sense for any entry is 2018. So we'll need to screen out entries with `registration_year` less than 1885, and greater than 2018.

In [17]:
autos = autos[autos['registration_year'].between(1885,2018)]
autos['registration_year'].value_counts(normalize=True).sort_index()

1910    0.000042
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
          ...   
2014    0.013774
2015    0.007895
2016    0.024378
2017    0.028787
2018    0.009827
Name: registration_year, Length: 80, dtype: float64

## 2 - Analyzing the data

In this section, we'll be interrogating the dataset, trying to hopefully answer some questions about the used car market in Germany.

### 2.1 - Average prices and mileage of car brands with the largest market shares in Germany

Suppose we wanted to know what the average prices of car brands with the largest market shares in Germany are. For this question, we assume that the percentage of listings a particular car brand has is a decent estimate of its market share in the German market.

Let's see what the data tells us.

In [18]:
market_share = autos['brand'].value_counts().sort_values(ascending=False)/autos.shape[0]
display(market_share)

volkswagen        0.212808
bmw               0.109816
opel              0.107003
mercedes_benz     0.096924
audi              0.086845
ford              0.068493
renault           0.047349
peugeot           0.029543
fiat              0.025491
seat              0.018856
skoda             0.016168
nissan            0.015286
mazda             0.015181
smart             0.014488
citroen           0.014215
toyota            0.012808
hyundai           0.010016
sonstige_autos    0.009386
volvo             0.009092
mini              0.008735
mitsubishi        0.008126
honda             0.008063
kia               0.007181
alfa_romeo        0.006635
porsche           0.005879
suzuki            0.005774
chevrolet         0.005690
chrysler          0.003528
dacia             0.002709
daihatsu          0.002499
jeep              0.002247
land_rover        0.002079
subaru            0.002058
saab              0.001638
daewoo            0.001554
jaguar            0.001491
rover             0.001344
t

Let's define "largest market shares" as car brands with >2.5% market share. We can screen for cars that meet this pre-requisite and aggregate their average price. We'll include the average odometer reading as well, as a bonus.

In [19]:
top_market_share_brands = market_share[market_share > 0.025].index
avg_price_popular_brands = {}
avg_mileage_popular_brands = {}
for brand in top_market_share_brands:
    sum_of_prices = autos.loc[autos['brand'] == brand, 'price_usd'].sum()
    sum_of_mileage = autos.loc[autos['brand'] == brand, 'odometer_km'].sum()
    nr_of_entries = autos.loc[autos['brand'] == brand, 'price_usd'].shape[0]
    
    avg_price = sum_of_prices/nr_of_entries
    avg_mileage = sum_of_mileage/nr_of_entries
    
    avg_price_popular_brands[brand] = avg_price
    avg_mileage_popular_brands[brand] = avg_mileage

price_series = pd.Series(avg_price_popular_brands)
mileage_series = pd.Series(avg_mileage_popular_brands)
agg_df = pd.DataFrame(price_series, columns = ['avg_price_usd'])
agg_df['avg_odometer_km'] = mileage_series
agg_df

avg_price_usd  avg_odometer_km
volkswagen       5434.265121    129025.160335
bmw              8330.403633    132919.694073
opel             3040.491758    129372.056515
mercedes_benz    8589.008232    131143.847487
audi             9283.795696    129620.406190
ford             3861.124157    124178.418148
renault          2503.220843    128044.345898
peugeot          3114.116560    127125.088842
fiat             2899.282537    117421.746293

The most popular car brand (volkswagen) has an average resell price of ~\\$5000. Somewhat surprisingly, the second most popular car brand is BMW has an average price of ~\\$8000. This could mean that a lot of folks choose to get their luxury cars (BMWs, Mercedes, etc) on the resale market over buying new at a dealership. I think this could be the case, seeing as 3 out of the top 5 resold brands are luxury brands. Additionally, I explored whether this noticeable price delta is due to a difference in the average mileages between car brands for sale. It looks like there is no noticeable difference between the average mileage between luxury brands and non-luxury brands, hence the price difference is not due to a mileage difference.

## 3 - Next steps

In this project, we clenaed and interrogated data from the used car market on eBay in Germany. Here's a couple of next steps we could take/questions we could answer to push this project forward:
* How has the popularity of stick shift cars changed over the years?
* What types of vehicles are sold in Germany and which is the most popular?
* Has the popularity of car ownership changed over the years?